lib/chef/rest.rb

```diff
@@ -39,6 +39,152 @@ class REST < HTTP; end
+      200 => 'OK',
+      201 => 'Created',
+      202 => 'Accepted',
+      203 => 'Non-Authoritative Information',
+      204 => 'No Content',
+      205 => 'Reset Content',
+      206 => 'Partial Content',
+      207 => 'Multi-Status',
+      300 => 'Multiple Choices',
+      301 => 'Moved Permanently',
+      302 => 'Found',
+      303 => 'See Other',
+      304 => 'Not Modified',
+      305 => 'Use Proxy',
+      307 => 'Temporary Redirect',
+      400 => 'Bad Request',
+      401 => 'Unauthorized',
+      402 => 'Payment Required',
+      403 => 'Forbidden',
+      404 => 'Not Found',
+      405 => 'Method Not Allowed',
+      406 => 'Not Acceptable',
+      407 => 'Proxy Authentication Required',
+      408 => 'Request Timeout',
+      409 => 'Conflict',
+      410 => 'Gone',
+      411 => 'Length Required',
+      412 => 'Precondition Failed',
+      413 => 'Request Entity Too Large',
+      414 => 'Request-URI Too Large',
+      415 => 'Unsupported Media Type',
+      416 => 'Request Range Not Satisfiable',
+      417 => 'Expectation Failed',
+      422 => 'Unprocessable Entity',
+      423 => 'Locked',
+      424 => 'Failed Dependency',
+      426 => 'Upgrade Required',
+      428 => 'Precondition Required',
+      429 => 'Too Many Requests',
+      431 => 'Request Header Fields Too Large',
+      500 => 'Internal Server Error',
+      501 => 'Not Implemented',
+      502 => 'Bad Gateway',
+      503 => 'Service Unavailable',
+      504 => 'Gateway Timeout',
+      505 => 'HTTP Version Not Supported',
+      507 => 'Insufficient Storage',
+      511 => 'Network Authentication Required',
+    }
+
+    STATUS_MESSAGE.values.each {|v| v.freeze }
+    STATUS_MESSAGE.freeze
+
+    def initialize(base_url)
+      @url = base_url
+    end
+
+    def host
+      @url.hostname
+    end
+
+    def port
+      @url.port
+    end
+
+    # request, response = client.request(method, url, body, headers) {|r| r.read_body }
+    def request(method, url, body, headers, &handler_block)
+      #pp req: [method, url, body, headers]
+      body_str = body || ""
+      r = {}
+      r["REQUEST_METHOD"] = method.to_s.upcase
+      r["SCRIPT_NAME"] = ""
+      r["PATH_INFO"] = url.path
+      r["QUERY_STRING"] = url.query
+      r["SERVER_NAME"] = "localhost"
+      r["SERVER_PORT"] = url.port
+      r["HTTP_HOST"] = "localhost:#{url.port}"
+      r["rack.url_scheme"] = "chefzero"
+      r["rack.input"] = StringIO.new(body_str)
+
+      res = ChefZero::SocketlessServerMap.request(port, r)
+
+      net_http_response = to_net_http(res[0], res[1], res[2])
+
+      yield net_http_response if block_given?
+
+      [self, net_http_response]
+    end
+
+    # TODO: this is copied verbatim from the fakeweb project, MIT licensed
+    # Add credits where appropriate
+
+    def to_net_http(code, headers, chunked_body)
+      body = chunked_body.join('')
+      msg = STATUS_MESSAGE[code] or raise "Cannot determine HTTP status message for code #{code}"
```

Comment:
This case should be rare, but when it happens, it'll be helpful to have the rest of the response in the exception message so you can work out what the request was.